### ISSUE TITLE
paged results

### ISSUE THREAD

**Issue by doowb** [jonschlinkert/github-base#18]
From [this PR in gists](https://github.com/jonschlinkert/gists/pull/3), I started thinking about how `.paged` attempts to return all results from an endpoint. This may cause requests to hit rate limits and result in errors.

I think we can make `.paged` return an object (similar to a javascript iterator) that has a `.next()` method on it. This will handle setting up the next request using the "next link" header information from the previous request and executing the request to return the results. The results can be an object that conforms to [the iterator protocols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols). I think we can implement it so `.next` can take a callback so we can use the api in older versions of node.

User code might look something like this (using `async/await`):

```js
async function getRepos(user) {
  let github = new GithubBase();
  let results = [];

  // get repos iterator
  let repos = github.paged('/users/:user/repos', {user: user});

  // get first page of results
  let res = await repos.next();

  // as long as there are more results, concat them to the results array and get the next page
  while (!res.done) {
    results = results.concat(res.value.repos);
    // handle "sleeping" due to github rate limits
    await ratelimit(res.value.response);

    // get the next page of results
    res = await repos.next();
  }
}

async function ratelimit(response) {
  // code to handle rate limit from response headers
}

let repos = await getRepos('doowb');
```

And by handling a callback in `.next`, as I mentioned above, we could do this with the node async pattern too.

**Comment by jonschlinkert** [comment]
love this idea @doowb!

**Comment by jonschlinkert** [comment]
I just [published this library](https://github.com/jonschlinkert/paged-request). I kept the code as simple and generalized as possible, but we can tweak if necessary.

**Comment by doowb** [comment]
Awesome library! I think that can be used here to achieve the type of functionality I was thinking. I really like the way you put the `pages` property on the `acc` object that's passed to the user-defined `next` function. This gives the user the ability to clear that list if necessary and use each page as it's coming through (or even to decide to return an undefined url to stop processing).

**Comment by jonschlinkert** [comment]
closing since this is resolved in 1.0
